api/http/management.go

```diff
@@ -15,38 +15,203 @@
-func NewManagementController(app app.App) *ManagementController {
-	return &ManagementController{
-		app: app,
+func NewManagementHandler(h *APIHandler, config ...*Config) *ManagementHandler {
+	conf := NewConfig(config...)
+	mh := &ManagementHandler{
+		APIHandler: h,
+	}
+	if conf.Client != nil {
+		mh.Client = conf.Client
+	} else {
+		mh.Client = new(http.Client)
+	}
+	return mh
+}
+
+func (h *ManagementHandler) proxyAzureRequest(c *gin.Context, dstPath string) {
+	req := c.Request
+	ctx := req.Context()
+	settings, err := h.app.GetSettings(ctx)
+	switch {
+	case err != nil:
+		_ = c.Error(err)
+		rest.RenderError(c,
+			http.StatusInternalServerError,
+			errors.New(http.StatusText(http.StatusInternalServerError)),
+		)
+		return
+	case settings.ConnectionString == nil:
+		fallthrough
+	case settings.ConnectionString.Validate() != nil:
+		rest.RenderError(c, http.StatusConflict, ErrMissingConnectionString)
+	default:
+	}
+	cs := settings.ConnectionString
+	q := c.Request.URL.Query()
+	q.Set("api-version", AzureAPIVersion)
+	req.URL.Scheme = "https"
+	req.URL.RawQuery = q.Encode()
+	req.Host = cs.HostName
+	req.URL.Path = dstPath
+
+	req.URL.Host = cs.HostName
+	if cs.GatewayHostName != "" {
+		req.URL.Host = cs.GatewayHostName
+	}
+	req.RequestURI = ""
+
+	delHbHHeaders(req.Header)
+
+	var expireAt time.Time
+	if dl, ok := ctx.Deadline(); ok {
+		expireAt = dl
+	} else {
+		var cancel context.CancelFunc
+		expireAt = time.Now().Add(defaultTTL)
+		ctx, cancel = context.WithDeadline(ctx, expireAt)
+		defer cancel()
+		req = req.WithContext(ctx)
+	}
+	req.Header.Set(HdrKeyAuthz, cs.Authorization(expireAt))
+	if req.Header.Get(HdrKeyXFF) == "" {
+		if host, _, err := net.SplitHostPort(req.RemoteAddr); err == nil {
+			req.Header.Set(HdrKeyXFF, host)
+		}
+	}
+	rsp, err := h.Do(req)
+	if err != nil {
+		_ = c.Error(err)
+		rest.RenderError(c,
+			http.StatusBadGateway,
+			errors.New("failed to proxy request to IoT Hub"),
+		)
+		return
+	}
+	defer rsp.Body.Close()
+	delHbHHeaders(rsp.Header)
+	delete(rsp.Header, HdrKeyMSRequestID)
+	rspHdrs := c.Writer.Header()
+	for k, v := range rsp.Header {
+		rspHdrs[k] = v
+	}
+	c.Status(rsp.StatusCode)
+	_, err = io.Copy(c.Writer, rsp.Body)
+	if err != nil {
+		_ = c.Error(err)
 	}
 }
 
+func (h *ManagementHandler) GetDeviceModules(c *gin.Context) {
+	h.proxyAzureRequest(c, AzureURIDeviceModules.URI(c.Param("id")))
+}
+
+func (h *ManagementHandler) GetDevice(c *gin.Context) {
+	h.proxyAzureRequest(c, AzureURIDevice.URI(c.Param("id")))
+}
+
+// GET /device/:id/twin
+func (h *ManagementHandler) GetDeviceTwin(c *gin.Context) {
+	h.proxyAzureRequest(c, AzureURIDeviceTwin.URI(c.Param("id")))
+}
+
+// PATCH /device/:id/twin
+func (h *ManagementHandler) UpdateDeviceTwin(c *gin.Context) {
+	var desiredProps struct {
+		Properties struct {
+			Desired map[string]interface{} `json:"desired"`
+		} `json:"properties"`
+	}
+	err := c.ShouldBindJSON(&desiredProps.Properties.Desired)
```

Comment:
This is smart, but we have an issue when you PUT the device twin for a device with tags. In these case, the tags in the Azure Device Twins are lost. PATCH preserves them, because of the different semantic. I suggest to add support for both properties and tags in the same end-point, and I'm sorry for the new request of reworking this end-point.

```
{"properties:" {"a": "b"}, "tags": {"c": "d"}}
```

What do you think?

Reply:
Are you sure it will clear the tags? It wasn't very clear reading [the docs](https://docs.microsoft.com/en-us/rest/api/iothub/service/devices/replace-twin). Strictly speaking the entire device representation should disappear according to these specs if it is a "proper" HTTP PUT endpoint.

Reply:
Nevermind, I tested it and this is indeed the case. Fixing it...